### PR TITLE
Updated transformations API guide

### PR DESCRIPTION
**Pull request by GlebKazantaev**
Change list:
* Updated library description
* Updated Disabling/Enabling specific transformations for plugin X section
* Moved code snippets to CPP files
* Added RTTI to transformation examples

### CODE REVIEW

File: docs/nGraph_DG/nGraphTransformation.md [hunk 1]
```diff
@@ -32,7 +33,7 @@ But if some transformation parts can potentially be reused in other transformati
 To decide where to store your transformation code, please follow these rules:
 
 1. If it is a plugin-specific transformation and cannot be reused by other plugins, keep source code inside plugin.
-2. If this transformation relates to the OpSetXToOpSetY conversion or it is common optimization, keep sources inside the transformation library.
+2. If this transformation relates to opset operation conversion or optimization then keep sources inside the transformation library.
 
 After you decide where to store your transformation code, you can start developing your own nGraph transformation.
```
Comment (ilyachur) [inline docs/nGraph_DG/nGraphTransformation.md:38]: Could you update a paragraph about ngraph function and add that nGraph function contains pointers to  `Parameter`, `Result` and `Sink`?

File: docs/nGraph_DG/nGraphTransformation.md [hunk 2]
```diff
@@ -14,10 +14,11 @@ Transformation library is independent from Inference Engine target library named
 and is located in the `inference-engine/src/transformations` directory.
 
 Transformations root directory contains two folders:
-* `ngraph_ops` - Contains legacy opset operations needed for nGraph to CNNNetwork conversion.
-> **NOTE**: This operation is prohibited inside new plugins until they are not moved to a separate directory with allowed operations.
+* `ngraph_ops` - Contains internal opset operations that are common for plugins.
 * `transformations` - Includes all transformations, utils, runtime info attributes, and pass managers.
-> **NOTE**: Do not use transformation that belongs to `ngraph::pass::ConvertOpSet1ToLegacy` transformations until they are not moved to a separate directory with allowed transformations.
+
+All internal operations and transformations located inside [Transformation Library](group__ie__transformation__api.html) are allowed for usage inside plugins.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:20]:
```suggestion
All internal operations and transformations located inside the [Transformation Library](group__ie__transformation__api.html) can be used inside plugins.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 3]
```diff
@@ -14,10 +14,11 @@ Transformation library is independent from Inference Engine target library named
...
+All legacy operations and transformations were moved to legacy library and not suggested to use.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:21]:
```suggestion
All legacy operations and transformations were moved to a legacy library and are not recommended to be used.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 4]
```diff
@@ -32,7 +33,7 @@ But if some transformation parts can potentially be reused in other transformati
 To decide where to store your transformation code, please follow these rules:
 
 1. If it is a plugin-specific transformation and cannot be reused by other plugins, keep source code inside plugin.
-2. If this transformation relates to the OpSetXToOpSetY conversion or it is common optimization, keep sources inside the transformation library.
+2. If this transformation relates to opset operation conversion or optimization then keep sources inside the transformation library.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:36]:
```suggestion
2. If this transformation relates to opset operation conversion or optimization, keep sources inside the transformation library.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 5]
```diff
@@ -87,14 +88,15 @@ To use `ngraph::pass::MatcherPass`, you need to complete these steps:
...
+Pattern is a single root `ngraph::Function`. But the only difference is that you do not need to create a function object, you just need to create and connect opset or special pattern operations.
+Then you need to take the last created operation and put it as a root of the pattern. This root node will be used as a root node in pattern matching.
 > **NOTE**: Any nodes in a pattern that have no consumers and are not registered as root will not be used in pattern matching. 
 
 @snippet example_ngraph_utils.cpp pattern:simple_example
 
 The `Parameter` operation in the example above has type and shape specified. These attributes are needed only to create Parameter operation class and will not be used in pattern matching.
 
-For instructions on how to match a pattern where `ShapeOf` takes any operation as an input, follow the [pattern matching](#pattern_matching) section.
+For more pattern examples please follow the [pattern matching](#pattern_matching) section.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:99]:
```suggestion
For more pattern examples, refer to the [pattern matching](#pattern_matching) section.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 6]
```diff
@@ -153,6 +155,8 @@ And then creates map from registered MatcherPasses. That helps to avoid addition
 
 ![graph_rewrite_efficient_search] 
 
+> **NOTE**: GraphRewrite execution algorithm can not set manually and depends only on root nodes registered inside MatcherPasses.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:158]:
```suggestion
> **NOTE**: GraphRewrite execution algorithm cannot be set manually and depends only on root nodes registered inside MatcherPasses.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 7]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
 
 ## Disabling/Enabling specific transformations for plugin X	 <a name="disabling_transformation"></a>
 
-This topic is mostly related to conversion to legacy opset and plugins that are based on CNNNetwork. But this mechanism still can be applied for other cases.
-Let's suppose that plugin X enabled the `opset3::StridedSlice` operation support and you want to disable the `ngraph::pass::ConvertStridedSliceToCrop` transformation for plugin X.
-To do this, you need to create a callback on plugin side and pass it to transformation. And also you need to update particular transformation to use this callback.  
+In transformation library we offer to plugins transformations like CommonOptimizations that contains predefined sequence of transformations.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:406]:
```suggestion
In transformation library, we provide plugins transformations like CommonOptimizations, which contains predefined sequence of transformations.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 8]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
+So we also provide a tool that helps to disable or partially disable particular transformations in transformation pipeline.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:407]:
```suggestion
We also provide a tool that helps to disable or partially disable particular transformations in a transformation pipeline.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 9]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
+For example, if plugin uses CommonOptimization transformation and want to disable ConvertGELU transformation then inside plugin we have to take PassConfig instance
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:408]:
```suggestion
For example, if a plugin uses the CommonOptimization transformation and needs to disable the ConvertGELU transformation, then inside the plugin we have to take the PassConfig instance
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 10]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
-// Register transformation and pass callback to pass::Manager
-ngraph::pass::Manager manager;
-manager.register_pass<ngraph::pass::ConvertStridedSliceToCrop>();
-// pass::Manager will set callback to all reistered transformations automatically
-manager.set_callback(transformations_callback);
-manager.run_passes(f);
-```
+@snippet example_ngraph_utils.cpp ngraph:disable_gelu
+
+In some cases we want to disable transformation for some condition:
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:413]:
```suggestion
In some cases, we need to disable transformation for some condition:
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 11]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
+
+@snippet example_ngraph_utils.cpp ngraph:disable_callback
+
+And in some cases pass::Manager pipelines inside transformations may have transformations disabled by default but enabled insde plugins.
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:417]:
```suggestion
In some cases, pass::Manager pipelines inside transformations may have transformations disabled by default but enabled inside plugins.
```

File: docs/nGraph_DG/nGraphTransformation.md [hunk 12]
```diff
@@ -399,33 +403,22 @@ NGRAPH_ENABLE_VISUALIZE_TRACING=1 -  enables visualization after each transforma
...
+
+@snippet example_ngraph_utils.cpp ngraph:disabled_by_default
+
+PassConfig instance taken from pass::Manager is shared across all registered transformations including nested transformations. So it doesn't matter where we work with this object (before passes registration or after).
```
Comment (avladimi) [inline docs/nGraph_DG/nGraphTransformation.md:421]:
```suggestion
PassConfig instance taken from pass::Manager is shared across all registered transformations including nested transformations. So it does not matter where we work with this object (before passes registration or after).
```

File: docs/nGraph_DG/nGraphTransformation.md (hunk 1 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:38]: Actually I want to divide this doc into several parts. In this PR I juts update transformations API related topics.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 2 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:20]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 4 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:36]: done

File: docs/nGraph_DG/nGraphTransformation.md (hunk 6 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:158]: done

File: docs/nGraph_DG/nGraphTransformation.md (hunk 3 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:21]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 5 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:99]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 8 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:407]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 9 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:408]: done

File: docs/nGraph_DG/nGraphTransformation.md (hunk 10 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:413]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 11 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:417]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 12 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:421]: done.

File: docs/nGraph_DG/nGraphTransformation.md (hunk 7 above)
Comment (GlebKazantaev) [reply docs/nGraph_DG/nGraphTransformation.md:406]: done.

**Review by avladimi — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
Need to port to master as well.

**Pull request merged**